File: org.lflang/src/org/lflang/ErrorReporter.java
package org.lflang;

import java.nio.file.Path;

import org.eclipse.emf.ecore.EObject;
import org.eclipse.lsp4j.DiagnosticSeverity;

import org.lflang.generator.Position;

/**
 * Interface for reporting errors.
 *
 * @author Edward A. Lee
 * @author Marten Lohstroh
 * @author Christian Menard
 */
public interface ErrorReporter {

    /**
     * Report an error.
     *
     * @param message The error message.
     * @return a string that describes the error.
     */
    String reportError(String message);


    /**
     * Report a warning.
     *
     * @param message The warning message.
     * @return a string that describes the warning.
     */
    String reportWarning(String message);

    /**
     * Report an informational message.
     *
     * @param message The message to report
     * @return a string that describes the error
     */
    String reportInfo(String message);


    /**
     * Report an error on the specified parse tree object.
     *
     * @param object  The parse tree object.
     * @param message The error message.
     * @return a string that describes the error.
     */
    String reportError(EObject object, String message);


    /**
     * Report a warning on the specified parse tree object.
     *
     * @param object  The parse tree object.
     * @param message The error message.
     * @return a string that describes the warning.
     */
    String reportWarning(EObject object, String message);

    /**
     * Report an informational message on the specified parse tree object.
     *
     * @param object The parse tree object.
     * @param message The informational message
     * @return a string that describes the info
     */
    String reportInfo(EObject object, String message);


    /**
     * Report an error at the specified line within a file.
     *
     * @param message The error message.
     * @param line    The one-based line number to report at.
     * @param file    The file to report at.
     * @return a string that describes the error.
     */
    String reportError(Path file, Integer line, String message);


    /**
     * Report a warning at the specified line within a file.
     *
     * @param message The error message.
     * @param line    The one-based line number to report at.
     * @param file    The file to report at.
     * @return a string that describes the warning.
     */
    String reportWarning(Path file, Integer line, String message);


    /**
     * Report an informational message at the specified line within a file.
     *
     * @param file The file to report at.
     * @param line The one-based line number to report at.
     * @param message The error message.
     * @return
     */
    String reportInfo(Path file, Integer line, String message);

    /**
     * Report a message of severity {@code severity}.
     * @param file The file to which the message pertains, or {@code null} if the file is unknown.
     * @param severity the severity of the message
     * @param message the message to send to the IDE
     * @return a string that describes the diagnostic
     */
    default String report(Path file, DiagnosticSeverity severity, String message) {
        switch (severity) {
        case Error:
            return reportError(message);
        case Warning:
        case Hint:
        case Information:
            return reportInfo(message);
        default:
            return reportWarning(message);
        }
    }

    /**
     * Report a message of severity {@code severity} that
     * pertains to line {@code line} of an LF source file.
     * @param file The file to which the message pertains, or {@code null} if the file is unknown.
     * @param severity the severity of the message
     * @param message the message to send to the IDE
     * @param line the one-based line number associated
     *             with the message
     * @return a string that describes the diagnostic
     */
    default String report(Path file, DiagnosticSeverity severity, String message, int line) {
        switch (severity) {
        case Error:
            return reportError(file, line, message);
        case Warning:
        case Hint:
        case Information:
            return reportInfo(file, line, message);
        default:
            return reportWarning(file, line, message);
        }
    }

    /**
     * Report a message of severity {@code severity} that
     * pertains to the range [{@code startPos}, {@code endPos})
     * of an LF source file.
     * @param file The file to which the message pertains, or {@code null} if the file is unknown.
     * @param severity the severity of the message
     * @param message the message to send to the IDE
     * @param startPos the position of the first character
     *                 of the range of interest
     * @param endPos the position immediately AFTER the
     *               final character of the range of
     *               interest
     * @return a string that describes the diagnostic
     */
    default String report(Path file, DiagnosticSeverity severity, String message, Position startPos, Position endPos) {
        return report(file, severity, message, startPos.getOneBasedLine());
    }

    /**
     * Check if errors where reported.
     *
     * @return true if errors where reported
     */
    boolean getErrorsOccurred();
}

File: org.lflang/src/org/lflang/generator/ErrorReporter.kt
package org.lflang.generator

import org.eclipse.core.resources.IMarker
import org.eclipse.core.resources.IResource
import org.eclipse.emf.ecore.EObject
import org.eclipse.emf.ecore.resource.Resource
import org.eclipse.xtext.nodemodel.util.NodeModelUtils
import org.lflang.generator.cpp.CppFileConfig

class ErrorReporter(private val fileConfig: CppFileConfig) : IErrorReporter {

    private var errorsOccurred = false

    // TODO move somewhere else
    private fun Resource.toPath() = org.lflang.FileConfig.toPath(this)

    private val EObject.node get() = NodeModelUtils.getNode(this)

    /**
     * Report a warning or error on the specified object
     *
     * The caller should not throw an exception so execution can continue.
     * This will print the error message to stderr.
     * If running in INTEGRATED mode (within the Eclipse IDE), then this also
     * adds a marker to the editor.
     * @param message The error message.
     * @param severity One of IMarker.SEVERITY_ERROR or IMarker.SEVERITY_WARNING
     * @param obj The Ecore object, or null if it is not known.
     */
    private fun report(message: String, severity: Int, obj: EObject): String {
        val line = obj.node.startLine
        val resource = fileConfig.getIResource(obj.eResource())
        return report(message, severity, line, resource)
    }

    /**
     * Report a warning or error on the specified line of the specified resource.
     *
     * The caller should not throw an exception so execution can continue.
     * This will print the error message to stderr.
     * If running in INTEGRATED mode (within the Eclipse IDE), then this also
     * adds a marker to the editor.
     * @param message The error message.
     * @param severity One of IMarker.SEVERITY_ERROR or IMarker.SEVERITY_WARNING
     * @param line The line number or null if it is not known.
     * @param resource The resource, or null if it is not known.
     */
    override fun report(
        message: String,
        severity: Int,
        line: Int?,
        resource: IResource?,
    ): String {
        val isError = severity == IMarker.SEVERITY_ERROR
        if (isError) {
            errorsOccurred = true
        }

        val header = if (isError) "ERROR" else "WARNING"
        val fullPath = resource?.fullPath?.toString() ?: "unknown path"

        if (line == null || fullPath == "unknown path")
            System.err.println("$header:\n$message")
        else
            System.err.println("$header: $fullPath $line\n$message")

        /* TODO!
        // If running in INTEGRATED mode, create a marker in the IDE for the error.
        // See: https://help.eclipse.org/2020-03/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2FresAdv_markers.htm
        if (mode === Mode.INTEGRATED) {
            var myResource = resource
            if (myResource === null && object !== null) {
                // Attempt to identify the IResource from the object.
                val eResource = object.eResource
                if (eResource !== null) {
                    val uri = FileConfig.toPath(eResource).toUri();
                    myResource = getEclipseResource(uri);
                }
            }
            // If the resource is still null, use the resource associated with
            // the top-level file.
            if (myResource === null) {
                myResource = fileConfig.iResource
            }
            if (myResource !== null) {
                val marker = myResource.createMarker(IMarker.PROBLEM)
                marker.setAttribute(IMarker.MESSAGE, message);
                if (line !== null) {
                    marker.setAttribute(IMarker.LINE_NUMBER, line);
                } else {
                    marker.setAttribute(IMarker.LINE_NUMBER, 1);
                }
                // Human-readable line number information.
                marker.setAttribute(IMarker.LOCATION, lineAsString);
                // Mark as an error or warning.
                marker.setAttribute(IMarker.SEVERITY, severity);
                marker.setAttribute(IMarker.PRIORITY, IMarker.PRIORITY_HIGH);

                marker.setAttribute(IMarker.USER_EDITABLE, false);

                // NOTE: It might be useful to set a start and end.
                // marker.setAttribute(IMarker.CHAR_START, 0);
                // marker.setAttribute(IMarker.CHAR_END, 5);
            }
        }*/

        // Return a string that can be inserted into the generated code.
        return "[[$header: $message]]"
    }

    /**
     * Report an error.
     * @param message The error message.
     */
    override fun reportError(message: String): String = report(message, IMarker.SEVERITY_ERROR, null, null)

    /**
     * Report a warning.
     * @param message The warning message.
     */
    override fun reportWarning(message: String): String = report(message, IMarker.SEVERITY_WARNING, null, null)

    /**
     * Report an error on the receiving parse tree object.
     *
     * @param obj     The parse tree object.
     * @param message The error message.
     */
    override fun reportError(obj: EObject, message: String): String = report(message, IMarker.SEVERITY_ERROR, obj)

    /**
     * Report a warning on the receiving parse tree object.
     *
     * @param obj     The parse tree object.
     * @param message The error message.
     */
    override fun reportWaring(obj: EObject, message: String): String = report(message, IMarker.SEVERITY_WARNING, obj)

    /** Report an error on the receiving parse tree object.
     *  @receiver The parse tree object.
     *  @param message The error message.
     */
    fun EObject.extReportError(message: String): String = report(message, IMarker.SEVERITY_ERROR, obj = this)

    /** Report a warning on the receiving parse tree object.
     *  @receiver The parse tree object.
     *  @param message The error message.
     */
    fun EObject.extReportWarning(message: String): String = report(message, IMarker.SEVERITY_WARNING, obj = this)
}